packages/ra-keycloak/src/authProvider.ts

```diff
@@ -103,55 +77,147 @@ export type PermissionsFunction = (decoded: KeycloakTokenParsed) => any;
         });
     },
     async logout() {
-        return client.logout({
-            redirectUri: options.logoutRedirectUri ?? window.location.origin,
-        });
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        return keycloakClient
+            .logout({
+                redirectUri:
+                    options.logoutRedirectUri ??
+                    `${window.location.origin}/login`,
+            })
+            .catch(error => {
+                console.error(error);
+            });
     },
     async checkError() {
-        return Promise.resolve();
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (!keycloakClient.authenticated || !keycloakClient.token) {
+            throw new Error('Failed to obtain access token.');
+        }
     },
     async checkAuth() {
-        return client.authenticated && client.token
-            ? Promise.resolve()
-            : Promise.reject('Failed to obtain access token.');
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (keycloakClient.authenticated && keycloakClient.token) {
+            return;
+        }
+        // not authenticated: save the location that the user tried to access
+        localStorage.setItem(
+            PreviousLocationStorageKey,
+            window.location.href.replace(window.location.origin, '')
+        );
+        throw new Error('Failed to obtain access token.');
     },
     async getPermissions() {
-        if (!client.token) {
-            return Promise.resolve(false);
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (!keycloakClient.token) {
+            return false;
         }
-        const decoded = jwt_decode<KeycloakTokenParsed>(client.token);
-        return Promise.resolve(
-            options.onPermissions ? options.onPermissions(decoded) : decoded
-        );
+        const decoded = jwt_decode<KeycloakTokenParsed>(keycloakClient.token);
+        return options.onPermissions ? options.onPermissions(decoded) : decoded;
     },
     async getIdentity() {
-        if (client.token) {
-            const decoded = jwt_decode<KeycloakTokenParsed>(client.token);
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (keycloakClient.token) {
+            const decoded = jwt_decode<KeycloakTokenParsed>(
+                keycloakClient.token
+            );
             const id = decoded.sub || '';
             const fullName = decoded.preferred_username;
-            return Promise.resolve({ id, fullName });
+            return { id, fullName };
+        }
+        throw new Error('Failed to get identity.');
+    },
+    async handleCallback(params: any) {
```

Comment:
nitpick: `params` is unused

```suggestion
    async handleCallback() {
```